brave/browser/net/proxy/proxy_config_service_tor.cc

```diff
@@ -0,0 +1,162 @@
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "base/path_service.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_util.h"
+#include "chrome/common/chrome_paths.h"
+#include "crypto/random.h"
+#include "net/proxy/proxy_service.h"
+#include "net/url_request/url_request_context.h"
+#include "services/service_manager/public/cpp/connector.h"
+
+namespace net {
+
+using content::BrowserThread;
+const int kTorPasswordLength = 16;
+
+ProxyConfigServiceTor::ProxyConfigServiceTor(const std::string tor_path,
+  const std::string tor_proxy) {
+    if (tor_path.length() && tor_proxy.length()) {
+      tor_path_ = tor_path;
+      url::Parsed url;
+      url::ParseStandardURL(
+        tor_proxy.c_str(),
+        std::min(tor_proxy.size(),
+                 static_cast<size_t>(std::numeric_limits<int>::max())),
+        &url);
+      if (url.scheme.is_valid()) {
+        scheme_ =
+          std::string(tor_proxy.begin() + url.scheme.begin,
+                      tor_proxy.begin() + url.scheme.begin + url.scheme.len);
+      }
+      if (url.host.is_valid()) {
+        host_ =
+          std::string(tor_proxy.begin() + url.host.begin,
+                      tor_proxy.begin() + url.host.begin + url.host.len);
+      }
+      if (url.port.is_valid()) {
+        port_ =
+          std::string(tor_proxy.begin() + url.port.begin,
+                      tor_proxy.begin() + url.port.begin + url.port.len);
+      }
+
+      content::ServiceManagerConnection::GetForProcess()
+        ->GetConnector()
+        ->BindInterface(tor::mojom::kTorServiceName,
+                        &tor_launcher_);
+
+      tor_launcher_.set_connection_error_handler(
+        base::BindOnce(&ProxyConfigServiceTor::OnTorLauncherCrashed,
+                       base::Unretained(this)));
+
+      BrowserThread::PostTask(
+          BrowserThread::PROCESS_LAUNCHER, FROM_HERE,
+          base::Bind(&ProxyConfigServiceTor::LaunchTorProcess,
+                     base::Unretained(this)));
+    }
+    config_.proxy_rules().ParseFromString(std::string(scheme_ + "://" + host_
+      + ":" + port_));
+}
+
+void ProxyConfigServiceTor::OnTorCrashed(int32_t pid) {
+  LOG(ERROR) << "Tor Process(" << pid << ") Crashed";
+  BrowserThread::PostTask(
+      BrowserThread::PROCESS_LAUNCHER, FROM_HERE,
+      base::Bind(&ProxyConfigServiceTor::LaunchTorProcess,
+                 base::Unretained(this)));
+}
+
+ProxyConfigServiceTor::~ProxyConfigServiceTor() {}
+
+void ProxyConfigServiceTor::LaunchTorProcess() {
+  std::vector<std::string> args;
+  args.push_back("--ignore-missing-torrc");
+  args.push_back("-f");
+  args.push_back("/nonexistent");
+  args.push_back("--defaults-torrc");
+  args.push_back("/nonexistent");
+  args.push_back("--SocksPort");
+  args.push_back(host_ + ":" + port_);
+  base::FilePath user_data_dir;
+  PathService::Get(chrome::DIR_USER_DATA, &user_data_dir);
+  if (!user_data_dir.empty()) {
```

Comment:
Can we fail noisily in the case that the user_data_dir is unset?  It seems imprudent to let tor pick somewhere to scribble outside ~/.config/brave if it doesn't exist.  Not a big issue; if this is not convenient then don't worry about it.

Reply:
It is unlikely to happen and even if it does bunch of browser functions will not work

Reply:
the browser won't launch at all if user_data_dir is not set